### PR TITLE
Homotopy Continuation extension and remake

### PR DESCRIPTION
**Pull request by TorkelE**
With the introduction of package extensions (https://pkgdocs.julialang.org/dev/creating-packages/#Conditional-loading-of-code-in-packages-(Extensions)) we can add some additional support for packages like HomotopyContinuation without making incompatibilities messier.

This introduces a single function, `hc_steady_states` which finds the steady states of a reaction system using homotopy continuation. If initial conditions are provided, conservation laws are automatically utilised. This feature is stored in a new "HomotopyContinuationExtension" extension, and only available if HomoyopyContinuation is loaded.

This is the full docstring of the new function:
    `hc_steady_states(rs::ReactionSystem, ps; filter_negative=true, neg_thres=-1e-20, u0=typeof(ps)(), kwargs...)`

Uses homotopy continuation to find the steady states of the ODE system corresponding to the provided reaction system.

Arguments:
- `rs::ReactionSystem`: The reaction system for which we want to find the steady states.
- `ps`: The parameter values for which we want to find the steady states.
- `filter_negative=true`: If set to true, solutions with any species concentration <0 is removed from the output.
- `neg_thres=-1e-20`: Determine the minimum values for which a species concentration is to be considred non-negative. Species conentrations ``> neg_thres`` but `< 0.0` are set to `0.0`.
- `u0=typeof(ps)()`: Initial conditions for which we want to find the steady states. For systems with conservation laws this are required to compute conserved quantities.
- `kwargs...`: any additional arguments (like `show_progress= true`) are passed into HomotopyContinuation.jl's `solve` call. 

Examples:
```@repl
rs = @reaction_network begin
    k1, Y --> 2X
    k2, 2X --> X + Y
    k3, X + Y --> Y
    k4, X --> 0
end
ps = [:k3 => 1.0, :k2 => 2.0, :k4 => 1.5, :k1=>8.0]
hc_sol = hc_steady_states(rs, ps)
```
gives
```
[0.5000000000000002, 2.0000000000000004]
[0.0, 0.0]
[4.499999999999999, 5.999999999999999]
```

Notes:
- This is a wrapper around the `solve` function provided by HomotopyContinuation.jl, all credit for this functionality to that package's authors.

### CODE REVIEW

**Comment by ChrisRackauckas** [comment]
I do think that the specialization here should be able to be removed when it becomes a standard NonlinearSolve solver. Though I don't know what the time line on that will be.

**Comment by TorkelE** [comment]
When that happens we shoudl be able to just delete the extension, but in the meantime this should be good.

**Comment by isaacsas** [comment]
CI is still failing?

**Comment by isaacsas** [comment]
A test failed.

**Comment by TorkelE** [comment]
You mean this one:
image
it is for Julia 1.6 though, right?

I will try to check what is going on and return back, but if it is failing on 1.6 and not the latest version (if my interpretation is correct), it might be a mess to sort out, especially if the problem is in HC. Either case, will try to figure out what it is.

**Comment by isaacsas** [comment]
SciML is still supporting 1.6 till the next LTS release, so we need to figure out the issue. If it is HC we should open an issue for them to fix it.

**Comment by TorkelE** [comment]
ok

**Comment by TorkelE** [comment]
I think I know it. 1.6 does not support extension, hence this not working. This kind of defeats the entire purpose of this PR. Ideally I would go ahead anyway, but https://pkgdocs.julialang.org/v1/creating-packages/#Transition-from-normal-dependency-to-extension seems like an OK workaround so will go with that (hopefully being able to revert it at some point.

File: test/runtests.jl
Comment (isaacsas) [inline test/runtests.jl:51]: The test probably also needs a check that the extension mechanism exists to run conditionally then?

Comment (TorkelE) [reply test/runtests.jl:51]: I think this should work, given the work around (but should hold my horses until the 1.6 test actually passes).

However, an alternative would be to, for extensions, simply disable those tests on 1.6. I will let you decide.

Currently, I added 
```julia
if !isdefined(Base, :get_extension) # Workaround for 1.6 compatibility.
    include("../ext/CatalystHomotopyContinuationExtension.jl")
end
```
to Catalyst.jl to make things work on 1.6.

Comment (TorkelE) [reply test/runtests.jl:51]: Indeed I spoke to quickly. Is disabling the test for this case as you mentioned a working solution?

Comment (isaacsas) [reply test/runtests.jl:51]: Yeah, and going forward we can just do the same for any future extensions.

File: docs/src/catalyst_applications/homotopy_continuation.md
Comment (isaacsas) [inline docs/src/catalyst_applications/homotopy_continuation.md:12]: Perhaps also mention the Catalyst citation?

File: docs/src/catalyst_applications/homotopy_continuation.md
Comment (isaacsas) [inline docs/src/catalyst_applications/homotopy_continuation.md:53]: I think we should still say what Catalyst is doing under the hood before this. You can say it is done automatically, but we should advertise this functionality is in Catalyst...

Comment (isaacsas) [reply docs/src/catalyst_applications/homotopy_continuation.md:53]: We want people to appreciate that this isn't something done by HC for them, but is a Catalyst feature / functionality that many packages are lacking.

File: docs/src/catalyst_applications/homotopy_continuation.md
Comment (isaacsas) [inline docs/src/catalyst_applications/homotopy_continuation.md:57]:
```suggestion
hc_steady_states(wilhelm_2009_model, ps; u0)

```

File: docs/src/catalyst_applications/homotopy_continuation.md
Comment (isaacsas) [inline docs/src/catalyst_applications/homotopy_continuation.md:61]:
```suggestion
- `hc_steady_states` supports any systems where all rates are systems of rational polynomials (such as Hill functions with integer Hill coefficients).

```

File: ext/CatalystHomotopyContinuationExtension/homotopy_continuation_extension.jl
Comment (isaacsas) [inline ext/CatalystHomotopyContinuationExtension/homotopy_continuation_extension.jl:11]: Isn't it then any species concentration < neg_thres?

File: ext/CatalystHomotopyContinuationExtension/homotopy_continuation_extension.jl
Comment (isaacsas) [inline ext/CatalystHomotopyContinuationExtension/homotopy_continuation_extension.jl:13]: Why assume `u0=typeof(ps)()`? Just make it `nothing` if nothing is passed in.

File: ext/CatalystHomotopyContinuationExtension/homotopy_continuation_extension.jl
Comment (isaacsas) [inline ext/CatalystHomotopyContinuationExtension/homotopy_continuation_extension.jl:35]:
```suggestion
```
This doesn't seem necessary...

File: ext/CatalystHomotopyContinuationExtension/homotopy_continuation_extension.jl
Comment (isaacsas) [inline ext/CatalystHomotopyContinuationExtension/homotopy_continuation_extension.jl:6]:
```suggestion
Uses homotopy continuation via HomotopyContinuation.jl to find the steady states of the ODE system corresponding to the provided reaction system.
```

File: HISTORY.md
Comment (isaacsas) [inline HISTORY.md:4]: Add an example for using it here.

File: ext/CatalystHomotopyContinuationExtension/homotopy_continuation_extension.jl
Comment (isaacsas) [inline ext/CatalystHomotopyContinuationExtension/homotopy_continuation_extension.jl:42]:
```suggestion
    return (filter_negative ? filter_negative_f(sols; neg_thres) : sols)
```

File: ext/CatalystHomotopyContinuationExtension/homotopy_continuation_extension.jl
Comment (isaacsas) [inline ext/CatalystHomotopyContinuationExtension/homotopy_continuation_extension.jl:48]: I don't think you want to promote every parameter to `Float64`, for the same reason it is bad for MTK to do so. Can't you just preserve the user's input type (i.e. if they give a mixed tuple keep using tuples until you substitute in the values, and then substitute in the same type as the user specified)?

File: ext/CatalystHomotopyContinuationExtension/homotopy_continuation_extension.jl
Comment (isaacsas) [inline ext/CatalystHomotopyContinuationExtension/homotopy_continuation_extension.jl:54]: I've kind of wondered for a while if we should just not register all these and let them be substituted in directly by Symbolics? i.e. if one does 
```julia
mm2(X,v,K) = v*X/(X+K)
rn2 = @reaction_network a begin
       mm2(X,v,K), A --> B
       end
equations(rn2)
```
one gets
```julia
julia> equations(rn2)
1-element Vector{Union{Equation, Reaction}}:
 (X*v) / (K + X), A --> B
```
I guess the disadvantage is we can't autodetect them if we want to know about their presence, but we don't do that currently anyways.

File: ext/CatalystHomotopyContinuationExtension/homotopy_continuation_extension.jl
Comment (isaacsas) [inline ext/CatalystHomotopyContinuationExtension/homotopy_continuation_extension.jl:61]:
```suggestion
    vars_with_vals = Set(p[1] for p in pre_varpmap) 
    union!(vars_with_vals, keys(ModelingToolkit.defaults(rs))
```

File: ext/CatalystHomotopyContinuationExtension/homotopy_continuation_extension.jl
Comment (isaacsas) [inline ext/CatalystHomotopyContinuationExtension/homotopy_continuation_extension.jl:62]:
```suggestion
    issubset(species(rs), vars_with_vals) && return
```

File: ext/CatalystHomotopyContinuationExtension/homotopy_continuation_extension.jl
Comment (isaacsas) [inline ext/CatalystHomotopyContinuationExtension/homotopy_continuation_extension.jl:64]:
```suggestion
    isempty(conservedequations(rs)) || 
        error("The system has conservation laws but initial conditions were not provided for all species.")
```

File: ext/CatalystHomotopyContinuationExtension/homotopy_continuation_extension.jl
Comment (isaacsas) [inline ext/CatalystHomotopyContinuationExtension/homotopy_continuation_extension.jl:86]: Calling invoke seems hacky. 

Perhaps we should just not register the functions to begin with? Was there any good reason we did so originally?

Comment (isaacsas) [reply ext/CatalystHomotopyContinuationExtension/homotopy_continuation_extension.jl:86]: I guess such a change might need to wait for the next breaking release though.

File: ext/CatalystHomotopyContinuationExtension/homotopy_continuation_extension.jl
Comment (isaacsas) [inline ext/CatalystHomotopyContinuationExtension/homotopy_continuation_extension.jl:89]: I have no idea what this is doing, but why do you need to wrap and unwrap? Shouldn't equations already be unwrapped (and stay that way)?

File: src/Catalyst.jl
Comment (TorkelE) [reply src/Catalyst.jl:111]: I was a little bit uncertain how we wanted to do it, but agree with, and prefer to, not to have 1.6 support.

File: ext/CatalystHomotopyContinuationExtension/homotopy_continuation_extension.jl
Comment (isaacsas) [inline ext/CatalystHomotopyContinuationExtension/homotopy_continuation_extension.jl:94]: Shouldn't this give an error if there is a non-integer exponent, since such a model couldn't be converted to a polynomial form?

File: ext/CatalystHomotopyContinuationExtension/homotopy_continuation_extension.jl
Comment (TorkelE) [reply ext/CatalystHomotopyContinuationExtension/homotopy_continuation_extension.jl:86]: I and Shashi sat together and fixed this (well Shashi mostly). He was interested in improving this though, so we opened an issue to hopefully have some native support fro this in Symbolics: https://github.com/JuliaSymbolics/Symbolics.jl/issues/975

File: ext/CatalystHomotopyContinuationExtension/homotopy_continuation_extension.jl
Comment (TorkelE) [reply ext/CatalystHomotopyContinuationExtension/homotopy_continuation_extension.jl:89]: This was also part of Shashi's magic, we had some decent number of iterations to go reach here, so a couple of more natural solutions failed.

File: ext/CatalystHomotopyContinuationExtension/homotopy_continuation_extension.jl
Comment (TorkelE) [reply ext/CatalystHomotopyContinuationExtension/homotopy_continuation_extension.jl:94]: It errors eventually, but I agree that a proper error here is way better, will add.

File: ext/CatalystHomotopyContinuationExtension/homotopy_continuation_extension.jl
Comment (isaacsas) [inline ext/CatalystHomotopyContinuationExtension/homotopy_continuation_extension.jl:121]:
```suggestion
        (neg_thres < sol[idx] < 0) && (sol[idx] = 0)
```
Just use `0`, it is more general with regards to types.

File: ext/CatalystHomotopyContinuationExtension/homotopy_continuation_extension.jl
Comment (isaacsas) [inline ext/CatalystHomotopyContinuationExtension/homotopy_continuation_extension.jl:123]:
```suggestion
    return filter(sol -> all(>=(0), sol), sols)
```

File: ext/CatalystHomotopyContinuationExtension/homotopy_continuation_extension.jl
Comment (TorkelE) [reply ext/CatalystHomotopyContinuationExtension/homotopy_continuation_extension.jl:54]: wasn't there some advantage that we have recorded the various derivatives of `mm` etc. , so these are known? Maybe the yare calculated at instant speed anyway though. I have no strong opinions, and maybe we should open an issue to discuss the pros/cons (I will mostly defer to your opinions though)

File: ext/CatalystHomotopyContinuationExtension/homotopy_continuation_extension.jl
Comment (isaacsas) [inline ext/CatalystHomotopyContinuationExtension/homotopy_continuation_extension.jl:105]:
```suggestion
        return sum(remove_denominators(arg) for arg in arguments(s_expr))
```
Does this work instead?

**Review by isaacsas — CHANGES_REQUESTED** [review]
I think those are all my comments if you want to make updates.

File: ext/CatalystHomotopyContinuationExtension/homotopy_continuation_extension.jl
Comment (isaacsas) [reply ext/CatalystHomotopyContinuationExtension/homotopy_continuation_extension.jl:54]: I can't remember why we did this change either unfortunately.

Comment (isaacsas) [reply ext/CatalystHomotopyContinuationExtension/homotopy_continuation_extension.jl:54]: The only thought I have is we can programmatically check for hill equations via the function type, which Chris and I discussed adding a JumpProcesses specialization for.

File: src/Catalyst.jl
Comment (isaacsas) [reply src/Catalyst.jl:111]: Let's just make it 1.9+, and note that in the HISTORY file.

**Comment by TorkelE** [comment]
Added your changes, except for a few that feel more like long-term questions regarding symbolic function handling.  Also let `u0` default to `nothing` and didn't use the `subset` check regarding conservation laws, see comments.

File: src/miscellaneous_functions.jl
Comment (isaacsas) [inline src/miscellaneous_functions.jl:36]: You need to look at the real part of the eigenvalues as they can be complex. Also, a system may not be unstable if the real parts are all non-negative (i.e. some are zero), so we probably shouldn't return a boolean but an enum that indicates, stable, unstable, eigenvalues are non-negative with at least one zero.

Maybe this can be moved to a future PR when we can better work out how to do this (and so we can get this merged now)?

File: ext/CatalystHomotopyContinuationExtension/homotopy_continuation_extension.jl
Comment (isaacsas) [reply ext/CatalystHomotopyContinuationExtension/homotopy_continuation_extension.jl:61]: Why not use `issubset`? It avoids allocations, unlike using `intersect`...

File: ext/CatalystHomotopyContinuationExtension/homotopy_continuation_extension.jl
Comment (isaacsas) [inline ext/CatalystHomotopyContinuationExtension/homotopy_continuation_extension.jl:47]: Do you have a test that this all works if the user passes a tuple for the parameter map using mixed types like integers and floats (even if the u0 map is a vector)?

File: ext/CatalystHomotopyContinuationExtension/homotopy_continuation_extension.jl
Comment (isaacsas) [reply ext/CatalystHomotopyContinuationExtension/homotopy_continuation_extension.jl:61]: Also, don't you want to enforce that all species have an initial value?

**Comment by TorkelE** [comment]
- stability computation removed.
- Added Tuple with mixed int/float to tests.
- Now combines p/u0 using:
```julia
[symmap_to_varmap(rs,u0)..., symmap_to_varmap(rs,ps)...]
```
(I tried other stuff, but couldn't come up with something that worked well on this combination which didn't require more lines of code. Given what we are doing next, I presume that the performance damage will be relatively minimal, and I didn't want to reduce readability.
- Now uses this check:
```julia
any(s -> s in vars_with_vals, species(rs))
```
I think part of the problem with your suggestion was that you required initial conditions for *all* species, while I only want to require for some. This way when you have a system with many species and a single conserved quantity you don't have to provide tons of initial conditions. I have also pointed this out in docs properly.

**Pull request merged**
